### PR TITLE
Onboard explicit sharding to deepseek

### PR DESCRIPTION
**Pull request by NuojCheng**
# Description

New replacement PR: #2783 

In this PR we made following changes:
1. Add sharding constraints/reshards for explicit sharding bug-free;
2. Add attention tests on explicit sharding, including expert parallelism, context parallelism, and mixture of two
3. Refactor logical names/sharding constraints initialization in `attention.py`, `attention_mla.py`.

# Tests

Model_name: deepseek3-test
Topology: v5p-8
JAX==0.8.0
TL;DR: 
* Auto Sharding: Maintains performance parity with the main branch (deviations ≤ 0.3%).
* Explicit Sharding: While this introduces a slight expected overhead in standard cases, it provides significant improvements for TP_transpose sharding.

## Performance Impact Table

| Configuration | Main Peak Memory (Mb) | Main Step Time (ms) | Auto Peak Memory (Mb) | Auto Peak Mem Change (%) | Auto Step Time (ms) | Auto Step Time Change (%) | Explicit Peak Memory (Mb) | Explicit Peak Mem Change (%) | Explicit Step Time (ms) | Explicit Step Time Change (%) |
| :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- |
| **FSDP4** | 68748 | 1822 | 68748 | 0% | 1822 | 0% | 68961 | 0.31% | 1870 | 2.63% |
| **FSDP2+TP2** | 69447 | 1427 | 69445 | 0% | 1431 | 0.28% | 69616 | 0.24% | 1796 | 25.86% |
| **FSDP2+TPT2** | 93835 | 10382 | 93836 | 0% | 10385 | 0.03% | 69372 | -26.07% | 2179 | -79.01% |
| **FSDP2+CP2** | 69296 | 2340 | 69476 | 0.26% | 2340 | 0% | 69453 | 0.23% | 2395 | 2.35% |
| **FSDP2+EP2 (fsdp)** | 68653 | 1718 | 68656 | 0% | 1719 | 0.06% | 68846 | 0.28% | 1823 | 6.11% |
| **FSDP2+EP2 (context)** | 69240 | 2293 | 69240 | 0% | 2293 | 0% | 69263 | 0.03% | 2506 | 9.29% |

## Full FSDP
```
smoke_train model_name=deepseek3-test per_device_batch_size=1
```
* aca5b24931ebcbadb55a82e56ebffe8024874028: [xprof](http://xprof.corp.google.com/trace_viewer/chengnuojin-2762930717666032893)
* `shard_mode=auto`: [xprof](http://xprof.corp.google.com/trace_viewer/chengnuojin-8060230202672386049)
* `shard_mode=explicit`: [xprof](http://xprof.corp.google.com/trace_viewer/chengnuojin-5555350233668210700)

## FSDP + TP
```
smoke_train model_name=deepseek3-test per_device_batch_size=1 ici_tensor_parallelism=2
```

* aca5b24931ebcbadb55a82e56ebffe8024874028: [xprof](http://xprof.corp.google.com/trace_viewer/chengnuojin-8870811566292461778)
* `shard_mode=auto`: [xprof](http://xprof.corp.google.com/trace_viewer/chengnuojin-14060932592428206835)
* `shard_mode=explicit`: [xprof](http://xprof.corp.google.com/trace_viewer/chengnuojin-10995837577894754784)

## FSDP + TP_transpose
```
smoke_train model_name=deepseek3-test per_device_batch_size=1 ici_tensor_transpose_parallelism=2
```

* aca5b24931ebcbadb55a82e56ebffe8024874028: [xprof](http://xprof.corp.google.com/trace_viewer/chengnuojin-13759446056080431764)
* `shard_mode=auto`: [xprof](http://xprof.corp.google.com/trace_viewer/chengnuojin-15418979663026557696)
* `shard_mode=explicit`: [xprof](http://xprof.corp.google.com/trace_viewer/chengnuojin-16855119196646264764)

## FSDP + CP
```
smoke_train model_name=deepseek3-test per_device_batch_size=1 ici_tensor_context_parallelism=2 dataset_type=synthetic
```

* aca5b24931ebcbadb55a82e56ebffe8024874028: [xprof](http://xprof.corp.google.com/trace_viewer/chengnuojin-9255731548424718585)
* `shard_mode=auto`: [xprof](http://xprof.corp.google.com/trace_viewer/chengnuojin-9839605967880876997)
* `shard_mode=explicit`: [xprof](http://xprof.corp.google.com/trace_viewer/chengnuojin-1507129928280236868)

## FSDP + EP(fsdp)
```
smoke_train model_name=deepseek3-test per_device_batch_size=1 ici_tensor_expert_parallelism=2  expert_shard_attention_option=fsdp
```

* aca5b24931ebcbadb55a82e56ebffe8024874028: [xprof](http://xprof.corp.google.com/trace_viewer/chengnuojin-18158158172310609911)
* `shard_mode=auto`: [xprof](http://xprof.corp.google.com/trace_viewer/chengnuojin-4315548038592028282)
* `shard_mode=explicit`: [xprof](http://xprof.corp.google.com/trace_viewer/chengnuojin-11107036957401468294)

## FSDP + EP(context)
```
smoke_train model_name=deepseek3-test per_device_batch_size=1 ici_tensor_expert_parallelism=2  expert_shard_attention_option=context dataset_type=synthetic
```

* aca5b24931ebcbadb55a82e56ebffe8024874028: [xprof](http://xprof.corp.google.com/trace_viewer/chengnuojin-10407907063405480861)
* `shard_mode=auto`: [xprof](http://xprof.corp.google.com/trace_viewer/chengnuojin-11120888670265014278)
* `shard_mode=explicit`: [xprof](http://xprof.corp.google.com/trace_viewer/chengnuojin-9490452156210735506)

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

File: src/MaxText/layers/moe.py [hunk 1]
```diff
@@ -1173,6 +1255,15 @@ def wrapper(x, logits, pre_bias_logits, w0, w1, wo, w0_bias, w1_bias, wo_bias, r
...
+    gate_logits = self._maybe_shard_with_logical(gate_logits, self.logical_names.gate)
+    pre_bias_logits = self._maybe_shard_with_logical(pre_bias_logits, self.logical_names.pre_bias)
+    w0_kernel = self._maybe_shard_with_logical(w0_kernel, self.logical_names.wi_kernel_sp)
+    w1_kernel = self._maybe_shard_with_logical(w1_kernel, self.logical_names.wi_kernel_sp)
+    wo_kernel = self._maybe_shard_with_logical(wo_kernel, self.logical_names.wo_kernel_sp)
+    w0_bias = self._maybe_shard_with_logical(w0_bias, self.logical_names.wi_bias)
+    w1_bias = self._maybe_shard_with_logical(w1_bias, self.logical_names.wi_bias)
+    wo_bias = self._maybe_shard_with_logical(w1_bias, self.logical_names.wo_bias)
     return wrapper(
         inputs, gate_logits, pre_bias_logits, w0_kernel, w1_kernel, wo_kernel, w0_bias, w1_bias, wo_bias, self.rngs
```
[the remark this answers is not in the document]
Comment (NuojCheng) [reply src/MaxText/layers/moe.py:1321]: Fixed

File: tests/attention_test.py [hunk 2]
```diff
@@ -546,41 +548,95 @@ def tpu_kernel_attention_helper(self, num_kv_heads):
...
           "expert_shard_attention_option": "context",
+          "shard_mode": "auto",
+      },
+      {
+          "testcase_name": "cp_no_load_balance_explicit",
+          "ici_context_parallelism": 4,
+          "context_parallel_load_balance": False,
+          "ici_expert_parallelism": 1,
+          "expert_shard_attention_option": "fsdp",
+          "shard_mode": "explicit",
```
[the remark this answers is not in the document]
Comment (NuojCheng) [reply tests/attention_test.py:600]: Fixed.

File: src/MaxText/layers/moe.py [hunk 3]
```diff
@@ -237,6 +261,15 @@ def __call__(self, inputs: jax.Array, _initializing: bool = False) -> Tuple[jax.
     kernel = jnp.asarray(kernel, self.dtype)
 
     contract_ind = tuple(range(0, len(norm_axis)))
+
+    # [B, S, E] -> [B, S, num_exp]
```
Comment (RissyRan) [inline src/MaxText/layers/moe.py:265]: Why do we need to shard embedding dimension to be expert?

File: src/MaxText/layers/moe.py [hunk 4]
```diff
@@ -393,6 +424,108 @@ def __init__(
...
+    else:
+      input_axes = (batch_logical_axis, "activation_norm_length", None)
+      wi_bias_axes = ("exp", "activation_mlp")
+      wo_bias_axes = ("exp", "activation_embed")
+
+    gate_logits_axes = (batch_logical_axis, "activation_norm_length", None)
+    if self.config.model_name.startswith("deepseek3"):
+      pre_bias_logits_axes = (batch_logical_axis, "activation_norm_length", None)
+    else:
+      # pre_bias_logits is None for non-DeepSeek v3 models
```
Comment (RissyRan) [inline src/MaxText/layers/moe.py:468]: Nit: let remove this comment, as if/else already indicates this logic right?

File: src/MaxText/layers/moe.py [hunk 5]
```diff
@@ -393,6 +424,108 @@ def __init__(
...
+      pre_bias_logits_axes = None
+
+    # w0, w1, wo needs to be un sharded on fsdp / fsdp_transpose axis, so use
+    # mlp_no_fsdp axis
+    if self.config.fsdp_shard_on_exp:
+      # special sharding for dsv3
+      wi_kernel_axes = ("embed_no_exp", None, "mlp")
```
Comment (RissyRan) [inline src/MaxText/layers/moe.py:475]:
```
      wi_kernel_axes = ("embed_no_exp", None, "mlp")
      wo_kernel_axes = ("embed_no_exp", "mlp", None)
```
Can be moved to quantization block?

File: src/MaxText/layers/moe.py (hunk 4 above)
Comment (NuojCheng) [reply src/MaxText/layers/moe.py:468]: sg

File: src/MaxText/layers/moe.py (hunk 5 above)
Comment (NuojCheng) [reply src/MaxText/layers/moe.py:475]: I think `wi_kernel_axes` and `wi_kernel_axes_sp` are used for different purposes. The later one is only used for sparse_matmul. For non-quantized+dense_matmul we stil use `wi_kernel_axes`

File: src/MaxText/layers/moe.py (hunk 3 above)
Comment (NuojCheng) [reply src/MaxText/layers/moe.py:265]: The `output_sharding` here specifies the logical axis of the matmul output, which is supposed to be B x S x NumExpert

**Comment by NuojCheng** [comment]
Thank you Ran for the reviews! @RissyRan 

> 1. split this PR into onboarding explicit sharding and refactor?

I can do that.

> 2. I see you have tested auto-mode and explicit mode, which is great! Have you tested before/after change for auto mode? We don't need to test all shardings like you have done already, but just ensure no perf regression for auto mode for DS v3 with basic FSDP, TP, EP sharding.

The performance test results between main and `shard_mode=auto` are included in the description table, under the column of `Auto Peak Mem Change (%)` and `Auto Step Time Change (%)`.

File: src/MaxText/layers/attention_mla.py [hunk 6]
```diff
@@ -60,6 +65,20 @@
 from MaxText.layers.linears import DenseGeneral
 from MaxText.layers.normalizations import RMSNorm
 from MaxText.layers.quantizations import AqtQuantization as Quant
+from MaxText.sharding import maybe_shard_with_logical
+
+
+@dataclasses.dataclass(frozen=True)
+class AttentionMLALogicalNames:
```
Comment (gobbleturk) [inline src/MaxText/layers/attention_mla.py:72]: cc @richjames0

File: src/MaxText/layers/attention_mla.py [hunk 7]
```diff
@@ -94,26 +113,6 @@ def mla_as_linen(
     temperature_tuning: bool = False,
     temperature_tuning_scale: float = 0.1,
     temperature_tuning_floor_scale: float = 8192.0,
-    # Shard the query activation as the same as the key and value.
```
Comment (gobbleturk) [inline src/MaxText/layers/attention_mla.py:97]: <3

File: src/MaxText/layers/attentions.py [hunk 8]
```diff
@@ -537,6 +471,64 @@ def _init_projections(self, inputs_q_shape: Tuple, inputs_kv_shape: Tuple) -> No
       self.value = self.init_kv_w(inputs_kv_shape=inputs_kv_shape)
     self.out = self.init_out_w(output_dim=inputs_q_shape[-1])
 
+  def _create_sharding(self, axis_names):
```
Comment (gobbleturk) [inline src/MaxText/layers/attentions.py:474]: cc @richjames0

File: src/MaxText/layers/moe.py (hunk 3 above)
Comment (gobbleturk) [reply src/MaxText/layers/moe.py:265]: I would have to think more but I think we want to leave the NumExpert dimension unsharded. The batch dimension should still be sharded by EP here (batch is sharded by EP until after the a2a)

File: tests/attention_test.py [hunk 9]
```diff
@@ -1489,12 +1610,16 @@ def _forward_with_context_expert_parallelism(cfg_cp, mesh_cp, attention_cp, lnx,
     decoder_positions = reordered_batch["inputs_position"]
   # apply attention with sharding
   with mesh_cp, nn_partitioning.axis_rules(cfg_cp.logical_axis_rules):
+    if cfg_cp.expert_shard_attention_option == EP_AS_CONTEXT:
```
Comment (gobbleturk) [inline tests/attention_test.py:1613]: oof we generally prefer to avoid logic in tests but I think I understand why this is here. Note this is a case that hints we should be modifying the logical axis rules at config parsing time instead of at runtime

Comment (gobbleturk) [reply tests/attention_test.py:1613]: cc @richjames0

File: src/MaxText/layers/llama2.py [hunk 10]
```diff
@@ -131,10 +137,33 @@ def __init__(
...
+    mlp_axis_names = ("activation_batch", "activation_length", "activation_mlp")
+    return LlamaLogicalNames(
+        inputs=activation_axis_names,
+        mlp=mlp_axis_names,
+        out=activation_axis_names,
+    )
+
+  def setup_sharding(self, model_mode):
+    self.logical_names = self._get_logical_names(model_mode)
+    self.inputs_sharding = self._create_sharding(self.logical_names.inputs)
```
Comment (RissyRan) [inline src/MaxText/layers/llama2.py:160]: Why those sharding uses `_create_sharding` instead of `maybe_shard_with_logical`? Which seems to return `None` in the auto mode?

File: src/MaxText/sharding.py [hunk 11]
```diff
@@ -48,10 +50,42 @@ def maybe_shard_with_logical(inputs, logical_axes, mesh, shard_mode):
...
   return maybe_shard_with_name(inputs, named_sharding, shard_mode)
 
 
+def remove_size_one_mesh_axis(spec, mesh) -> P:
+  """
+  Removes mesh axes from a PartitionSpec (P) where the axis size is 1.
+
+  This is a common optimization to simplify sharding by excluding redundant axes.
+  Copied from jax._src.core:
+  https://github.com/jax-ml/jax/blob/db65f4548d10493613ee4155fd463a17ab5680ddc/jax/_src/core.py#L2139-L2148
```
Comment (RissyRan) [inline src/MaxText/sharding.py:65]: This link is not found? https://screenshot.googleplex.com/48Vou6KgByUuGkQ You could generate the permanent link instead.

File: src/MaxText/layers/moe.py (hunk 4 above)
Comment (RissyRan) [reply src/MaxText/layers/moe.py:468]: Still there?

File: src/MaxText/layers/moe.py (hunk 5 above)
Comment (RissyRan) [reply src/MaxText/layers/moe.py:475]: OK I was a little bit confused. If so, could we rename them to `wi_kernel_axes_sparse` and `wi_kernel_axes_dense` for readability when creating them in one place?

File: src/MaxText/layers/moe.py (hunk 3 above)
Comment (RissyRan) [reply src/MaxText/layers/moe.py:265]: Yes, this is inside of Gating. So NumExpert will be starting to be sharded after 1st A2A and before 2nd A2A.

File: src/MaxText/layers/llama2.py (hunk 10 above)
Comment (NuojCheng) [reply src/MaxText/layers/llama2.py:160]: This is some sharding-related refactoring components. Let's pause it for now until we have design doc and more discussion on it.

File: src/MaxText/sharding.py (hunk 11 above)
Comment (NuojCheng) [reply src/MaxText/sharding.py:65]: Thank you for catching it. This component is supposed to in another PR #2737 . I removed this part in the new PR #2783
